### PR TITLE
ARROW-4341: [C++] Refactor Primitive builders and BooleanBuilder to use TypedBufferBuilder<T>

### PR DESCRIPTION
**Pull request by bkietz**
This reduces code duplication.

### CODE REVIEW

**Comment by bkietz** [comment]
Hmmm, this is a performance regression.

Before:
```
BM_BuildPrimitiveArrayNoNulls/repeats:2_mean                 365772 us     365779 us          1     2.291GB/s
BM_BuildVectorNoNulls/repeats:2_mean                         555089 us     555097 us          1   922.362MB/s
BM_BuildBooleanArrayNoNulls/repeats:2_mean                   121138 us     121139 us          6   4.12748GB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_mean                     13406 us      13406 us         52   4.66192GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_mean         28445 us      28445 us         25   2.19714GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_mean                    26330 us      26331 us         27   4.74726GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_mean        27448 us      27448 us         25   2.27696GB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_mean               136 us        136 us       5063   561.917MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_mean           125 us        125 us       5606   611.482MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_mean              126 us        126 us       5546   605.348MB/s
BM_BuildStringDictionaryArray/repeats:2_mean                    317 us        317 us       2209   315.468MB/s
```

After:
```
BM_BuildPrimitiveArrayNoNulls/repeats:2_mean                 417623 us     417631 us          1   1.75068GB/s
BM_BuildVectorNoNulls/repeats:2_mean                         555594 us     555601 us          1   921.525MB/s
BM_BuildBooleanArrayNoNulls/repeats:2_mean                   730000 us     730006 us          1   701.364MB/s
BM_BuildAdaptiveIntNoNulls/repeats:2_mean                     13394 us      13394 us         52   4.66611GB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_mean         28451 us      28451 us         25    2.1967GB/s
BM_BuildAdaptiveUIntNoNulls/repeats:2_mean                    26312 us      26313 us         27   4.75047GB/s
BM_BuildAdaptiveUIntNoNullsScalarAppend/repeats:2_mean        27457 us      27457 us         25    2.2762GB/s
BM_BuildInt64DictionaryArrayRandom/repeats:2_mean               139 us        139 us       5040   549.399MB/s
BM_BuildInt64DictionaryArraySequential/repeats:2_mean           126 us        126 us       5570   606.613MB/s
BM_BuildInt64DictionaryArraySimilar/repeats:2_mean              129 us        129 us       5406   589.165MB/s
BM_BuildStringDictionaryArray/repeats:2_mean                    319 us        319 us       2193   313.216MB/s
```

File: cpp/src/arrow/array/builder_primitive.h [hunk 1]
```diff
@@ -115,8 +124,9 @@ class ARROW_EXPORT PrimitiveBuilder : public ArrayBuilder {
   Status AppendValues(ValuesIter values_begin, ValuesIter values_end) {
     int64_t length = static_cast<int64_t>(std::distance(values_begin, values_end));
     ARROW_RETURN_NOT_OK(Reserve(length));
-
-    std::copy(values_begin, values_end, raw_data_ + length_);
+    for (auto it = values_begin; it != values_end; ++it) {
+      data_builder_.UnsafeAppend(*it);
+    }
```
Comment (wesm) [inline cpp/src/arrow/array/builder_primitive.h]: This is fishy

File: cpp/src/arrow/array/builder_primitive.h [hunk 2]
```diff
@@ -137,8 +147,9 @@ class ARROW_EXPORT PrimitiveBuilder : public ArrayBuilder {
                   "version instead");
     int64_t length = static_cast<int64_t>(std::distance(values_begin, values_end));
     ARROW_RETURN_NOT_OK(Reserve(length));
-
-    std::copy(values_begin, values_end, raw_data_ + length_);
+    for (auto it = values_begin; it != values_end; ++it) {
+      data_builder_.UnsafeAppend(static_cast<value_type>(*it));
+    }
```
Comment (wesm) [inline cpp/src/arrow/array/builder_primitive.h]: Also fishy

File: cpp/src/arrow/array/builder_primitive.h [hunk 3]
```diff
@@ -154,8 +165,9 @@ class ARROW_EXPORT PrimitiveBuilder : public ArrayBuilder {
       ValuesIter values_begin, ValuesIter values_end, ValidIter valid_begin) {
     int64_t length = static_cast<int64_t>(std::distance(values_begin, values_end));
     ARROW_RETURN_NOT_OK(Reserve(length));
-
-    std::copy(values_begin, values_end, raw_data_ + length_);
+    for (auto it = values_begin; it != values_end; ++it) {
+      data_builder_.UnsafeAppend(*it);
+    }
```
Comment (wesm) [inline cpp/src/arrow/array/builder_primitive.h]: Ditto

File: cpp/src/arrow/buffer-builder.h [hunk 4]
```diff
@@ -298,6 +298,17 @@ class TypedBufferBuilder<bool> {
...
   }
 
+  template <typename Generator>
+  void UnsafeAppend(const int64_t num_elements, Generator&& gen) {
+    if (num_elements == 0) return;
+    internal::GenerateBitsUnrolled(mutable_data(), bit_length_, num_elements, [&] {
+      bool value = gen();
+      if (!value) ++false_count_;
+      return value;
+    });
```
Comment (wesm) [inline cpp/src/arrow/buffer-builder.h]: This is a lot more CPU instructions than what it was doing before AFAICT

**Comment by xhochy** [comment]
Rebased.

**Comment by wesm** [comment]
BooleanBuilder seems to be more than 4 times slower with this patch

after

```
BM_BuildBooleanArrayNoNulls/repeats:2                        789455 us     789468 us          1   648.538MB/s
```

before

```
BM_BuildBooleanArrayNoNulls/repeats:2                        172299 us     172302 us          4   2.90189GB/s
```

Based on this evidence the iterator based unrolled bit generator vectorizes poorly. I just pushed changes https://github.com/apache/arrow/pull/3575/commits/be8442f8574cdc7d6df6f55d8f71136e217be894 reverting some of these changes (but still using `TypedBufferBuilder<bool>`, so better than before), as well as an option to disable tracking the false count, which we don't care about. For the record, disabling the false count only improved the benchmark from 650 MB/s to 1 GB/s, far less than the original 3GB/s. The compiler just isn't generating very good code for the CPU

If these changes are acceptable I suggest that we merge this patch (once the build is passing) and leave this be for now.

**Comment by wesm** [comment]
With my changes the perf seems slightly improved from the baseline

```
BM_BuildBooleanArrayNoNulls/repeats:2                        163077 us     162609 us          4   3.07486GB/s
```

**Comment by wesm** [comment]
It seems that there are symbol visibility problems on Windows caused by the PrimitiveBuilder / NumericBuilder collapse, and valgrind problems additionally. I would say to limit how much additional time (say 2 hours or less) you spend on this patch as there's a lot of feature work and bug fixes in the 0.13 backlog.

**Comment by wesm** [comment]
I moved the entire `NumericBuilder<T>` implementation to the header file. I think this should make the linker problems go away. I don't expect an application to have a ton of instantiations of these templates so I don't know that it's worth stressing about code size for these until it can be demonstrated to be a problem. The size of `arrow-array-test` in a release build increased by about 3%.

**Comment by wesm** [comment]
I reverted some changes and think I fixed the valgrind / Windows / Python failures

**Comment by wesm** [comment]
Seems there are some correctness issues https://travis-ci.org/apache/arrow/jobs/493616072. @bkietz can you look at the Ruby failure? We should try to reproduce the failure as a C++ unit test

**Comment by bkietz** [comment]
@wesm I'll try to write a unit test which produces the same error

File: cpp/src/arrow/buffer-builder.h [hunk 5]
```diff
@@ -347,6 +347,9 @@ class TypedBufferBuilder<bool> {
   }
 
   Status Finish(std::shared_ptr<Buffer>* out, bool shrink_to_fit = true) {
+    // set bytes_builder_.size_ == byte size of data
+    bytes_builder_.UnsafeAdvance(BitUtil::BytesForBits(bit_length_) -
+                                 bytes_builder_.length());
```
Comment (wesm) [inline cpp/src/arrow/buffer-builder.h]: Do these bytes need to be zeroed?

**Comment by bkietz** [comment]
I can't reproduce https://travis-ci.org/apache/arrow/jobs/493915377#L2632 locally

File: cpp/src/arrow/buffer-builder.h (hunk 5 above)
Comment (bkietz) [reply cpp/src/arrow/buffer-builder.h]: `TypedBufferBuilder<bool>` doesn't utilize `bytes_builder_`'s append methods, which means the size of bytes_builder_ is unchanged.

I'm thinking it's not worthwhile to use `BufferBuilder` in this class, I'm going to refactor and instead just manage a buffer directly

Comment (wesm) [reply cpp/src/arrow/buffer-builder.h]: OK, please do that refactoring in a follow up PR and not this one =)

**Comment by wesm** [comment]
OK, I can take it from here then

**Comment by wesm** [comment]
Fixed the Ruby test failures and added a C++ unit test that failed. 

@xhochy I removed the `valid_bytes` argument from `AppendNulls`

https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_primitive.h#L55

If anyone was using this argument, it was not doing anything useful, unless the goal is to append zeroes. I'm not sure if this needs to go through a deprecation cycle. In our codebase it was only being used for internal purposes

**Comment by wesm** [comment]
+1

**Pull request closed without merging**